File: license-server/src/main/java/de/netzkronehd/license/model/LicenseModel.java
package de.netzkronehd.license.model;

import de.netzkronehd.license.listmode.ListMode;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import lombok.*;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Objects;

@Entity
@Getter
@Setter
@ToString
@RequiredArgsConstructor
@AllArgsConstructor
public class LicenseModel {

    @Id
    private String license;

    private String publisher;
    @Column(columnDefinition = "TEXT")
    private String notes;
    private boolean valid;
    private OffsetDateTime validUntil;
    private ListMode listMode;

    @ElementCollection
    private List<String> ipAddresses;

    @Override
    public final boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof LicenseModel that)) return false;
        return valid == that.valid && Objects.equals(license, that.license) && Objects.equals(publisher, that.publisher) && Objects.equals(notes, that.notes) && Objects.equals(validUntil, that.validUntil) && listMode == that.listMode && Objects.equals(ipAddresses, that.ipAddresses);
    }

    @Override
    public int hashCode() {
        int result = Objects.hashCode(license);
        result = 31 * result + Objects.hashCode(publisher);
        result = 31 * result + Objects.hashCode(notes);
        result = 31 * result + Boolean.hashCode(valid);
        result = 31 * result + Objects.hashCode(validUntil);
        result = 31 * result + Objects.hashCode(listMode);
        result = 31 * result + Objects.hashCode(ipAddresses);
        return result;
    }
}

File: license-server-rs/src/model/license_model.rs
use crate::model::list_mode;
use std::fmt::{Display, Formatter};

pub struct LicenseModel {
    pub license: String,
    pub publisher: String,
    pub notes: String,
    pub valid: bool,
    pub valid_until: chrono::DateTime<chrono::FixedOffset>,
    pub list_mode: list_mode::ListMode,
    pub ip_addresses: Vec<String>,
}

impl Display for LicenseModel {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "LicenseModel {{ license: {:?}, publisher: {:?}, notes: {:?}, valid: {:?}, valid_until: {:?}, list_mode: {:?}, ip_addresses: {:?} }}",
            self.license,
            self.publisher,
            self.notes,
            self.valid,
            self.valid_until,
            self.list_mode,
            self.ip_addresses
        )
    }
}

impl PartialEq for LicenseModel {
    fn eq(&self, other: &Self) -> bool {
        self.license == other.license
            && self.publisher == other.publisher
            && self.notes == other.notes
            && self.valid == other.valid
            && self.valid_until == other.valid_until
            && self.list_mode == other.list_mode
            && self.ip_addresses == other.ip_addresses
    }
}
